src/julia_internal.h

```diff
@@ -861,12 +862,26 @@ jl_method_t *jl_make_opaque_closure_method(jl_module_t *module, jl_value_t *name
     int nargs, jl_value_t *functionloc, jl_code_info_t *ci, int isva, int isinferred);
 JL_DLLEXPORT int jl_is_valid_oc_argtype(jl_tupletype_t *argt, jl_method_t *source);
 
-STATIC_INLINE int is_anonfn_typename(char *name)
+STATIC_INLINE int is_anonfn_typename(char *name) JL_NOTSAFEPOINT
 {
     if (name[0] != '#' || name[1] == '#')
         return 0;
     char *other = strrchr(name, '#');
-    return other > &name[1] && other[1] > '0' && other[1] <= '9';
+    return other > &name[1] && isdigit(other[1]);
+}
+
+// Returns true for typenames of anounymous functions that have been canonicalized (i.e.
+// we mangled the name of the outermost enclosing function in their name).
+STATIC_INLINE int is_canonicalized_anonfn_typename(char *name) JL_NOTSAFEPOINT
+{
+    char *delim = strchr(&name[1], '#');
+    if (delim == NULL)
+        return 0;
+    if (delim[1] != '#')
+        return 0;
+    if (!isdigit(delim[2]))
```

Comment:
This needs a bounds check

Reply:
Not sure if I can see why? (At least assuming the string is properly null-terminated -- i.e. well-formed). 

Walking through the cases here:

1. If first `if` check returns true, we return. If it doesn't then we can guarantee that `delim[1]` will be valid (i.e. in-bounds but possibly a null-termination character).
2. If the second `if` check returns true, we return. If it doesn't, then `delim[1]` is equal to `#` and we can guarantee `delim[2]` will be valid (i.e. in-bounds but possibly a null-termination character).

Reply:
Bump?

Reply:
ah, gotcha, yes that makes sense

Reply:
Cool. Thanks.